File: 2026-04-April-GFG-POTD/April 07 - Stable Marriage Problem/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> stableMarriage(vector<vector<int>>& men, vector<vector<int>>& women) {
        int n = men.size();

        vector<int> wifeOfMan(n, -1);
        vector<int> husbandOfWoman(n, -1);
        vector<int> nextProposal(n, 0);

        // rank[w][m] = preference rank of man m for woman w
        vector<vector<int>> rank(n, vector<int>(n));
        for (int w = 0; w < n; w++) {
            for (int i = 0; i < n; i++) {
                rank[w][women[w][i]] = i;
            }
        }

        queue<int> freeMen;
        for (int i = 0; i < n; i++) {
            freeMen.push(i);
        }

        while (!freeMen.empty()) {
            int m = freeMen.front();
            freeMen.pop();

            int w = men[m][nextProposal[m]];
            nextProposal[m]++;

            if (husbandOfWoman[w] == -1) {
                // Woman is free
                husbandOfWoman[w] = m;
                wifeOfMan[m] = w;
            } else {
                int currentMan = husbandOfWoman[w];

                if (rank[w][m] < rank[w][currentMan]) {
                    // Woman prefers new man
                    husbandOfWoman[w] = m;
                    wifeOfMan[m] = w;

                    // Old man becomes free
                    wifeOfMan[currentMan] = -1;
                    freeMen.push(currentMan);
                } else {
                    // Woman rejects proposal
                    freeMen.push(m);
                }
            }
        }

        return wifeOfMan;
    }
};

File: 2026-04-April-GFG-POTD/April 07 - Stable Marriage Problem/Solution.py
from collections import deque

class Solution:
    def stableMarriage(self, men, women):
        n = len(men)

        wifeOfMan = [-1] * n
        husbandOfWoman = [-1] * n
        nextProposal = [0] * n

        # rank[w][m] = preference rank of man m for woman w
        rank = [[0] * n for _ in range(n)]
        for w in range(n):
            for i in range(n):
                rank[w][women[w][i]] = i

        freeMen = deque(range(n))

        while freeMen:
            m = freeMen.popleft()

            w = men[m][nextProposal[m]]
            nextProposal[m] += 1

            if husbandOfWoman[w] == -1:
                # Woman is free
                husbandOfWoman[w] = m
                wifeOfMan[m] = w
            else:
                currentMan = husbandOfWoman[w]

                if rank[w][m] < rank[w][currentMan]:
                    # Woman prefers new man
                    husbandOfWoman[w] = m
                    wifeOfMan[m] = w

                    # Old man becomes free
                    wifeOfMan[currentMan] = -1
                    freeMen.append(currentMan)
                else:
                    # Woman rejects proposal
                    freeMen.append(m)

        return wifeOfMan
